File: app/src/main/java/nontachai/becomedev/noteapp/LoginActivity.java
package nontachai.becomedev.noteapp;

import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.Toast;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.tuyenmonkey.mkloader.MKLoader;

import org.w3c.dom.Text;

import es.dmoral.toasty.Toasty;

public class LoginActivity extends AppCompatActivity {

    private TextInputLayout inputEmail, inputPass;
    Button btnLogin;
    private CheckBox checkBox;

    private FirebaseAuth firebaseAuth;

    private AdView mAdView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        inputEmail = findViewById(R.id.input_log_email);
        inputPass = findViewById(R.id.input_log_password);
        btnLogin = findViewById(R.id.btn_log);
        checkBox = findViewById(R.id.checkbox);

        firebaseAuth = FirebaseAuth.getInstance();

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setIcon(R.drawable.ic_save);

        ///load data from SharedPreferences....
        SharedPreferences sharedPref = getSharedPreferences("setting", MODE_PRIVATE);
        String mEmail = sharedPref.getString("email", null);
        String mPass = sharedPref.getString("password", null);
        inputEmail.getEditText().setText(mEmail);
        inputPass.getEditText().setText(mPass);


        btnLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                String lemail = inputEmail.getEditText().getText().toString().trim();
                String lpass = inputPass.getEditText().getText().toString().trim();


                if (lemail.isEmpty()) {
                    inputEmail.getEditText().setError("Enter email!!");
                } else if (lpass.isEmpty()) {
                    inputPass.getEditText().setError("Enter password!!");
                } else {
                    logIn(lemail, lpass);
                }

                if (checkBox.isChecked()) {
                    SharedPreferences sharedPreferences = getSharedPreferences("setting", MODE_PRIVATE);
                    SharedPreferences.Editor editor = sharedPreferences.edit();
                    editor.putString("email", lemail);
                    editor.putString("password", lpass);
                    editor.commit();

                }
            }
        });

        ///admob
        MobileAds.initialize(this, "ca-app-pub-1787292132881960~8520030186");//app id
        AdView adView = new AdView(this);
        adView.setAdSize(AdSize.BANNER);
        adView.setAdUnitId("ca-app-pub-1787292132881960/5723735633");//unit id  ca-app-pub-3940256099942544/6300978111

        mAdView = findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder().build();
        mAdView.loadAd(adRequest);
        ////end admob





    }

    private void logIn(String lemail, String lpass) {

        final ProgressDialog progressDialog = new ProgressDialog(this);
        progressDialog.setMessage("Logging in , please wait....");
        progressDialog.show();


        firebaseAuth.signInWithEmailAndPassword(lemail, lpass).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                progressDialog.dismiss();
                if (task.isSuccessful()) {
                    Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                    startActivity(intent);
                    finish();
                    Toasty.success(getApplicationContext(), "Sign In Success!", Toast.LENGTH_SHORT, true).show();
                } else {
                    Toasty.error(getApplicationContext(), "ERROR:" + task.getException().getMessage(), Toast.LENGTH_SHORT, true).show();
                    Log.d("LoginActivity", task.getException().getMessage());
                }
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);

        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return true;
    }
}

File: app/src/main/java/nontachai/becomedev/noteapp/LoginActivity.kt
package nontachai.becomedev.noteapp

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.text.method.PasswordTransformationMethod
import android.util.Log
import android.view.View
import android.widget.Toast
import cn.pedant.SweetAlert.SweetAlertDialog
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.OnCompleteListener

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.database.FirebaseDatabase
import kotlinx.android.synthetic.main.activity_login.*

const val TAG = "logd"

class LoginActivity : AppCompatActivity() {
    @SuppressLint("CommitPrefEdits")

    private val RC_SIGN_IN = 9001
    lateinit var auth:FirebaseAuth

    lateinit var dialogLoginGoogle : SweetAlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        edt_password.transformationMethod = PasswordTransformationMethod()
        edt_reg_password.transformationMethod = PasswordTransformationMethod()

        txt_signup.setOnClickListener {
            constraintLayout_register.visibility = View.VISIBLE
            constraintLayout_login.visibility = View.GONE
        }
        btn_back_to_login.setOnClickListener {
            constraintLayout_register.visibility = View.GONE
            constraintLayout_login.visibility = View.VISIBLE
        }

        //load save user data
        val sharedPref = getSharedPreferences("login_setting", Context.MODE_PRIVATE)
        val email = sharedPref.getString("email", "")
        val pass = sharedPref.getString("password", "")
        edt_username.setText(email)
        edt_password.setText(pass)



        //event on click
        btn_login.setOnClickListener {
            val username = edt_username.text.toString().trim()
            val password = edt_password.text.toString().trim()

            when {
                username.isEmpty() -> {
                    edt_username.error = "username is empty."
                    edt_username.requestFocus()
                }
                password.isEmpty() -> {
                    edt_password.error = "password is empty."
                    edt_password.requestFocus()
                }
                else -> {

                    val sharedPreferences = getSharedPreferences("login_setting", Context.MODE_PRIVATE)
                    val editor = sharedPreferences.edit()
                    if (chk_remember.isChecked) {
                        editor.putString("email", username)
                        editor.putString("password", password)
                        editor.apply()
                    } else {
                        editor.putString("email", "")
                        editor.putString("password", "")
                        editor.apply()
                    }

                    login(username, password)
                }
            }
        }

        btn_register.setOnClickListener {
            val name = edt_reg_name.text.toString()
            val emailReg = edt_reg_email.text.toString()
            val password = edt_reg_password.text.toString()

            when {
                name.isEmpty() -> {
                    edt_reg_name.error = "name is empty"
                    edt_reg_name.requestFocus()
                }
                emailReg.isEmpty() -> {
                    edt_reg_email.error = "email is empty"
                    edt_reg_email.requestFocus()
                }
                password.isEmpty() -> {
                    edt_reg_password.error = "password is empty"
                    edt_reg_password.requestFocus()
                }
                else -> {
                    register(name, emailReg, password)
                }
            }
        }

        dialogLoginGoogle = SweetAlertDialog(this,SweetAlertDialog.PROGRESS_TYPE)
        auth = FirebaseAuth.getInstance()
        btn_login_google.setOnClickListener {
            loginWithGoogle()
        }

        //auto
//        val userID = auth.currentUser?.uid
//        if(userID != "null" && userID != null){
//            Log.d(TAG, "data not null $userID")
//            login(email,pass)
//        }else{
//            Log.d(TAG, "data null $userID")
//        }
        

    }

    private fun register(name: String, email: String, password: String) {
        val fUserDatabase = FirebaseDatabase.getInstance().reference.child("Users")
        val firebaseAuth = FirebaseAuth.getInstance()
        firebaseAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                Log.d(TAG, "Create ac success: ")
                fUserDatabase.child(firebaseAuth.currentUser!!.uid).child("basic").child("name").setValue(name)
                        .addOnCompleteListener {
                            if (it.isSuccessful) {
                                //success register
                                SweetAlertDialog(this, SweetAlertDialog.WARNING_TYPE).apply {
                                    titleText = "Register Success!!"
                                    contentText = "back to login page,to login."
                                    show()
                                }
                            } else {
                                //err
                                SweetAlertDialog(this, SweetAlertDialog.WARNING_TYPE).apply {
                                    titleText = "Something wrong!!"
                                    contentText = "${task.exception!!.message}"
                                    show()
                                }
                            }
                        }
            } else {
                SweetAlertDialog(this, SweetAlertDialog.WARNING_TYPE).apply {
                    titleText = "Something wrong!!"
                    contentText = "${task.exception!!.message}"
                    show()
                }
            }
        }


    }

    private fun login(email: String, password: String) {
        val dialogLoading = SweetAlertDialog(this, SweetAlertDialog.PROGRESS_TYPE).apply {
            contentText = "Loading..."
            setCancelable(false)
            show()
        }

        val firebaseAuth = FirebaseAuth.getInstance()
        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener {
            if (it.isSuccessful) {
                //login success
                Log.d(TAG, ":login success ")
                val intent = Intent(this, MainActivity::class.java)
                startActivity(intent)
                dialogLoading.dismissWithAnimation()
            } else {
                //login fails
                Log.d(TAG, ":login failed ")
                dialogLoading.dismissWithAnimation()
                SweetAlertDialog(this, SweetAlertDialog.ERROR_TYPE).apply {
                    titleText = "Username/Password Incorrect!!"
                    contentText = "can't login please check your username/password."
                    show()
                }
            }
        }

    }


    private fun loginWithGoogle() {

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build()
        val googleSignInClient = GoogleSignIn.getClient(this, gso)
        val signInIntent = googleSignInClient.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)

    }
    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == RC_SIGN_IN) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            try {
                // Google Sign In was successful, authenticate with Firebase
                val account = task.getResult(ApiException::class.java)
                firebaseAuthWithGoogle(account!!)
            } catch (e: ApiException) {
                // Google Sign In failed, update UI appropriately
                Log.w(TAG, "Google sign in failed", e)
            }
        }
    }
    private fun firebaseAuthWithGoogle(acct: GoogleSignInAccount) {
        Log.d(TAG, "firebaseAuthWithGoogle:" + acct.id!!)

        dialogLoginGoogle.apply {
            contentText = "Loading..."
            setCancelable(false)
            show()
        }

        val credential = GoogleAuthProvider.getCredential(acct.idToken, null)
        auth.signInWithCredential(credential)
                .addOnCompleteListener(this) { task ->
                    if (task.isSuccessful) {
                        // Sign in success, update UI with the signed-in user's information
                        Log.d(TAG, "signInWithCredential:success")
                        //val user = auth.currentUser

                        dialogLoginGoogle.dismissWithAnimation()
                        val intent = Intent(this, MainActivity::class.java)

                        startActivity(intent)
                    } else {
                        // If sign in fails, display a message to the user.
                        Log.w(TAG, "signInWithCredential:failure", task.exception)
                    }
                }
    }

}
